File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        
        if not root:
            return result
        
        stack = [root]
        
        while stack:
            node = stack.pop()

            while node.left:
                temp = node.left
                node.left = None
                stack.append(node)
                node = temp

            result.append(node.val)
            if node.right:
                stack.append(node.right)
        return result

File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> lst = new ArrayList<>();
        helper(root, lst);
        return lst;
    }
    
    private void helper(TreeNode root, List<Integer> lst) {
        if (root == null) {
            return;
        }
        
        helper(root.left, lst);
        lst.add(root.val);
        helper(root.right, lst);
    }
}